Feature/transmission matrix

ph-schaefer opened the pull request:
## Feature: New class representing transmission matrices

A transmission matrix (T-matrix) is used to model a two-port or quadrupole, which can represent electrical circuits, mechanical systems (e.g. loudspeakers), acoustic transmission lines, and so on. Systems can be cascaded by multiplying matrices. Functions are provided to calculate the input and output impedance as well as transfer functions.

### Extension towards acoustic transmission lines and other applications
My personal goal is to create a class / module that allows to handle and cascade different type of transmission line segments, which are internally represented by T-matrices. Generally, other applications are possible (e.g. loudspeaker modeling), but would be beyone the scope of my implementation.
- Should such a representation be part of pyfar or a new package? -> not pyfar
- Could be a new pyfar module? -> for now personal module

### Numpy 1.24 (Python 3.8) and logical indicing
Numpy >= 1.25 does not support Python 3.8 anymore (see [here](https://endoflife.date/numpy)). I had some issues with numpy 1.24 (latest Version for Python 3.8) regarding logical indicing. I found a work around but generally logical indicing seems to a little more tricky in older versions.
For higher Python Versions, numpy 2.x was used and my original code worked. I did not check higher numpy 1.XX versions.

### Notes
- A change of the internal `_arithmetic()` method of the `_Audio` class was required

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+`Two-ports` (or `Quadrupoles`). These can represent systems from various fields,
+e.g. electrical circuits, mechanical vibration, (acoustic) transmission lines.
+
+Systems can be cascaded my multiplying consecutive T-matrices. Furthermore,
```
Maybe add a short example here to highlight how easy it is to use?
```suggestion
Systems can be cascaded my multiplying consecutive T-matrices (e.g., by simply using the ``@`` operator).
```

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+        shape = np.shape(data)
+        n_dim = len(shape)
+        if n_dim < 3 or shape[-3] != 2 or shape[-2] != 2:
+            raise ValueError("'data' must have a shape like [..., 2, 2, N]"
+                             ", e.g. [2, 2, 100].")
```
```suggestion
            raise ValueError("'data' must have a shape like (..., 2, 2, n_bins)"
                             ", e.g. (2, 2, 100).")
```

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+        ):
+            raise ValueError("A-,B-,C- and D-Matrices must be of the same type.")
+
+        if isinstance(A, FrequencyData):
+            frequencies = A.frequencies
+            A = A.freq
+            B = B.freq
+            C = C.freq
+            D = D.freq
+        if frequencies is None:
+            raise ValueError("'frequencies' must be specified if not using "
+                             "'FrequencyData' objects as input.")
+
+        data = np.array([[A, B], [C, D]])
```
I think this could be replaced by a combination of https://pyfar.readthedocs.io/en/stable/modules/pyfar.utils.html#pyfar.utils.concatenate_channels and https://pyfar.readthedocs.io/en/stable/classes/pyfar.audio.html#pyfar.classes.audio.FrequencyData.reshape, but I don't know it that's necessary.

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+    def abcd_cshape(self):
+        """The channel shape of the transmission matrix entries (A, B, C, D)
+
+        This is the same as 'cshape' without the last two elements.
```
```suggestion
        This is the same as 'cshape' without the last two elements (2, 2).
```

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+    @property
+    def A(self) -> FrequencyData:
+        """Returns the (potentially multi-dimensional) A entry of the T-matrix."""
+        return self[..., 0, 0, :]
```
Why not self[0, 0]?

ph-schaefer replied:
Fixed now

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+        """
+        return (self.D * Zl + self.B) / (self.C * Zl + self.A)
+
+    def transfer_function_quantity1(self, Zl: complex | FrequencyData):
```
I found them in l. 298 :D

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+
+    @classmethod
+    def _create_abcd_matrix_broadcast(cls, abcd_data: FrequencyData,
+                                      abcd_cshape : tuple):
```
Maybe add a little documentation for developers?

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+
+    @classmethod
+    def create_identity(cls, frequencies, abcd_cshape = ()):
+        """Creates an object with identity matrix entries (bypass).
```
Maybe add math equation for illustration?

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+    def create_series_impedance(cls, impedance : FrequencyData, abcd_cshape = ()):
+        """Creates a transmission matrix representing a series impedance.
+
+        This means the impedance is connected in series with a potential load impedance.
```
Maybe add math equation for illustration?

ph-schaefer replied:
Done

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,171 @@
+    _expect_data_with_wrong_abcd_dims(np.ones([7,8,4,2, len(frequencies)]), frequencies)
+
+
+def _get_example_init_data():
+    frequencies = [100, 200]
+    Mat_list = [1, 1]
+    Mat_np = np.array(Mat_list)
+    Mat_pf = FrequencyData(Mat_np, frequencies)
+    return frequencies, Mat_list, Mat_np, Mat_pf
```
Maybe create a fixture? https://pyfar-gallery.readthedocs.io/en/latest/contribute/contribution_guidelines.html#fixtures

sikersten commented on tests/test_transmission_matrix_derived_parameters.py:
```diff
@@ -0,0 +1,71 @@
+    npt.assert_allclose(TF_current.freq, 1/2, atol=1e-15)
+
+
+if __name__ == "__main__":
+    test_tmatrix_twoport_impedances("input")
+    test_tmatrix_twoport_impedances("output")
+    test_tmatrix_transfer_functions()
```
To be pushed to the repository?

sikersten commented on tests/test_transmission_matrix_derived_parameters.py:
```diff
@@ -0,0 +1,71 @@
+def test_tmatrix_twoport_impedances(type : str):
+    frequencies = [100,200,300]
+    rng = np.random.default_rng()
+    Zl = FrequencyData(rng.random(len(frequencies)), frequencies)
```
Does it make sense to explicitely test the cases with Zl=np.inf and Zl = 0?

sikersten reviewed (changes requested):
Excellent work, thanks @ph-schaefer! All in all, I only have minor comments.

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+                             "be of the same type.")
+
+        if isinstance(A, FrequencyData):
+            frequencies = A.frequencies
+            A = A.freq
+            B = B.freq
+            C = C.freq
+            D = D.freq
```
```suggestion
            frequencies = A.frequencies
            for d in (B,C,D):
                if isinstance(d, FrequencyData):
                    assert npt.allclose(
                        A.frequencies, d.frequencies, atol=1e-15)
                else:
                    raise ValueError(
                        "If using FrequencyData objects for A, "
                        "B, C, D, all must be FrequencyData objects.")
    
```
I think we need to make sure that all object have the same Frequency bins.

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+            FrequencyData type for A, B, C, D.
+
+        """
+        if (
+            not isinstance(B, type(A))
+            or not isinstance(C, type(A))
+            or not isinstance(D, type(A))
+        ):
+            raise ValueError("A-,B-,C- and D-Matrices must "
+                             "be of the same type.")
```
```suggestion
```
this is coverd by the changes below

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+        if frequencies is None:
+            raise ValueError("'frequencies' must be specified if not using "
+                             "'FrequencyData' objects as input.")
+
```
```suggestion
        # broadcast shapes
        shape = np.broadcast_shapes(
            np.array(A).shape, np.array(B).shape,
            np.array(C).shape, np.array(D).shape)
        A = np.broadcast_to(A, shape)
        B = np.broadcast_to(B, shape)
```
we should broadcast them to the same shape, in this way ``TransmissionMatrix.from_abcd(1,2,3,[5,6], frequencies)`` would be possible as well. FrequencyData are are automaticaly converted into arrays by np.array(object). So we dont need to take care of it. And numpy will take care of potential problems within the different arrays, e.g. type..
For us its simple, we just follow numpy here and let numpy do the job.

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+            A TransmissionMatrix object that contains 2x2 identity matrices.
+
+        """
+        identity = np.array([[1,0],[0,1]])
+        identity_3d = np.repeat(identity[:, :, np.newaxis],
+                                len(frequencies), axis = 2)
+        return cls._create_abcd_matrix_broadcast(
+            FrequencyData(identity_3d, frequencies), abcd_cshape)
```
```suggestion
        return cls.from_abcd(
            np.ones((*abcd_cshape, np.array(frequencies).size)), 0, 0, 1)
```
broadcasting magic

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+        if impedance.cshape != (1,):
+            raise ValueError("Number of channels for 'impedance' must be 1.")
+        Z = impedance.freq[0]
+        ones, zeros = np.ones_like(Z), np.zeros_like(Z)
+        series_Z_abcd = FrequencyData([[ones, Z],[zeros, ones]],
+                                      impedance.frequencies)
+        return cls._create_abcd_matrix_broadcast(series_Z_abcd, abcd_cshape)
```
```suggestion
        return cls.from_abcd(
            1, np.zeros((*abcd_cshape, np.array(frequencies).size))+Z,
            0, 1, impedance.frequencies)
```

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+        if admittance.cshape != (1,):
+            raise ValueError("Number of channels for 'admittance' must be 1.")
+        Y = admittance.freq[0]
+        ones, zeros = np.ones_like(Y), np.zeros_like(Y)
+        series_Z_abcd = FrequencyData([[ones, zeros],[Y, ones]],
+                                      admittance.frequencies)
+        return cls._create_abcd_matrix_broadcast(series_Z_abcd, abcd_cshape)
```
```suggestion        
        return cls.from_abcd(
            1, 0,
            np.zeros((*abcd_cshape, np.array(frequencies).size))+Y, 1, admittance.frequencies)
```

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+        Returns
+        -------
+        tmat : np.ndarray
+            A 2x2 array representing the transmission matrix of the transformer
+            and can be cascaded with TransmissionMatrix objects.
```
Doesn't it need to be of shape (2,2,1) to be multiplied by FrequencyData objects?

ph-schaefer replied:
No, since this is just an np.array and not a FrequencyData object, this is the correct approach.

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,610 @@
+        return nominator / denominator
+
+    @classmethod
+    def _create_abcd_matrix_broadcast(cls, abcd_data: FrequencyData,
+                                      abcd_cshape : tuple):
```
this would not be nessesaray if we apply broadcasting by definition, see my suggestions.

ahms5 commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,208 @@
+# -------------------------
+# TESTS FOR HIGHER DIM DATA
+# -------------------------
+@pytest.fixture(scope="module")
+def abcd_data_3x2():
+    """ABCD matrices with 2 frequency bins andone additional
+    dimension of size 3"""
+    frequencies = [100, 200]
+    A = FrequencyData([[1, 1], [1, 1], [1, 1]], frequencies)
+    B = FrequencyData([[2, 2], [2, 2], [2, 2]], frequencies)
+    C = FrequencyData([[3, 3], [3, 3], [3, 3]], frequencies)
+    D = FrequencyData([[4, 4], [4, 4], [4, 4]], frequencies)
+    tmat = TransmissionMatrix.from_abcd(A, B, C, D)
+    return tmat, A, B, C, D
+@pytest.fixture(scope="module")
+def abcd_data_3x3x1():
+    """ABCD matrices with 1 frequency bin and two additional
+    dimensions of size 3"""
+    A = FrequencyData(
+        [[[1.1], [1.1], [1.1]], [[1.2], [1.2], [1.2]], [[1.3], [1.3], [1.3]]],
+        100)
+    B = A + 1
+    C = A + 2
+    D = A + 3
+    tmat = TransmissionMatrix.from_abcd(A, B, C, D)
+    return tmat, A, B, C, D
```
this can be moved into conftests

ph-schaefer replied:
Since this is only used in one file/module, contest would the wrong place. See this discussion:
https://stackoverflow.com/questions/34466027/what-is-conftest-py-for-in-pytest

ahms5 reviewed (changes requested):
Thank your for implementing! looks fine, I have some suggestions for the array handling and i would include broadcasting by default. I think there are also some comments from simon which are not adressed yet.

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+        return self @ t_matrix
+
+    @property
+    def abcd_axes(self):
+        """The indices of the axes referring to the transmission matrix
+
+        This relates to the axes with respect to the full data set
+        (including the frequency-axis).
+
+        """
+        return (-3, -2)
```
This is probably not required and can be removed

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,395 @@
+        Parameters
+        ----------
+        data : array, double
+            Raw data in the frequency domain. The memory layout of Data is 'C'.
```
Probably not. I copied this from the audio classes documentation.

ph-schaefer commented on pyfar/classes/audio.py:
```diff
@@ -1354,6 +1354,8 @@ def matrix_multiplication(
 
 def _arithmetic(data: tuple, domain: str, operation: Callable, **kwargs):
     """Apply arithmetic operations."""
+    #NOTE: The import is done here to avoid a circular import
+    from pyfar.classes.transmission_matrix import TransmissionMatrix
```
This will probably lead to another circular import. This seems to be the best solution for now.

ahms5 reviewed (changes requested):
Thank you for your effort. I just have small comments, I think its almost ready

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,669 @@
+>>> tmat_out.freq == tmat.freq
+
+"""
+from __future__ import annotations
```
I think, this is required if you use your class type within the same file, e.g. as expected input or output type of a function. I must have needed this in an older version of the code, now it does not seem to be required anymore.

ph-schaefer replied:
It seems that without this, errors occur for Python <= 3.9, so I kept this line

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,669 @@
+        >>> import pyfar as pf
+        >>> frequencies = (100,200,300)
+
+        >>> # From np.array
```
I think this was fine. It is meant as a comment in a code block.

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,669 @@
+        >>> (B,C,D) = (A+1, A+2, A+3)
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
+
+        >>> # From FrequencyData objects
```
I think this was fine. It is meant as a comment in a code block.

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,669 @@
+        >>> (B,C,D) = (A+1, A+2, A+3)
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D)
+
+        >>> # Data with higher abcd dimension
```
I think this was fine. It is meant as a comment in a code block.

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        >>> A = np.ones(len(frequencies))
+        >>> (B,C,D) = (A+1, A+2, A+3)
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
+
```
```suggestion
    >>> tmat
    FrequencyData:
    (2, 2) channels with 3 frequencies
```
FrequencyData should be TransmissionMatrix, so we need to overwrite the ``FrequencyData.__repr__``

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        From FrequencyData objects
+        >>> A = pf.FrequencyData( A, frequencies )
+        >>> (B,C,D) = (A+1, A+2, A+3)
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D)
```
```suggestion
        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D)
        >>> tmat
        TransmissionMatrix:
        (2, 2) channels with 3 frequencies
```

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        Data with higher abcd dimension
+        >>> A = np.ones( (3, 4, len(frequencies)) )
+        >>> (B,C,D) = (A+1, A+2, A+3)
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
```
see above

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+            A FrequencyData object with the resulting output impedance. The
+            shape is identical to the entries of the T-matrix, i.e.
+            shape(tmat.A) == shape(Zout).
+
```
here we could add an example code

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+            A FrequencyData object with the resulting output impedance. The
+            shape is identical to the entries of the T-matrix, i.e.
+            shape(tmat.A) == shape(Zout).
+
```
here we could also add an example

ahms5 commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        shape = np.shape(data)
+        n_dim = len(shape)
+        if n_dim < 3 or shape[-3] != 2 or shape[-2] != 2:
+            raise ValueError("'data' must have a shape like "
```
not sure if we shouldnt stick to at least shape (1, 2, 2, N) so the cshape is at least (1, )
and what happen if i slice a TransmissionMatrix. At the moment its always a FrequencyData Object but shouldnt this example be a tMatrix if it keeps its shape (2,2)
```
frequencies = (100,200,300)
A = np.ones((1, len(frequencies)))
(B,C,D) = (A+1, A+2, A+3)
tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
type(tmat)  # pyfar.classes.transmission_matrix.TransmissionMatrix
type(tmat[0])  # pyfar.classes.audio.FrequencyData
```

ahms5 reviewed (changes requested):
Thanks for your quick application of the reviews! See the comments, I'm wondering about the slicing and representation of the class.

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        if impedance.cshape != (1,):
+            raise ValueError("Number of channels for 'impedance' must be 1.")
+
+        (Z, frequencies) = (impedance.freq[0], impedance.frequencies)
+        Z = np.zeros((*abcd_cshape, np.array(frequencies).size)) + Z
```
Is it possible to work with https://numpy.org/doc/stable/reference/generated/numpy.broadcast_to.html instead?

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        if admittance.cshape != (1,):
+            raise ValueError("Number of channels for 'admittance' must be 1.")
+
+        (Y, frequencies) = (admittance.freq[0], admittance.frequencies)
+        Y = np.zeros((*abcd_cshape, np.array(frequencies).size)) + Y
```
Possible to work with https://numpy.org/doc/stable/reference/generated/numpy.broadcast_to.html instead?

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,193 @@
+def frequencies():
+    return [100, 200]
+@pytest.fixture(scope="module")
+def Mat_list():
+    return [1, 1]
+@pytest.fixture(scope="module")
+def Mat_np(Mat_list):
+    return np.array(Mat_list)
+@pytest.fixture(scope="module")
+def Mat_pf(Mat_np, frequencies):
+    return FrequencyData(Mat_np, frequencies)
```
Could you add one-line comments to mention their purpose, please?

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,193 @@
+        ValueError, match=
+                    "If using FrequencyData objects, all matrix entries "
+                    "A, B, C, D, must be FrequencyData objects."):
+        TransmissionMatrix.from_abcd(A, B, C, D, 1000)
+def test_tmatrix_from_abcd_input_types(frequencies, Mat_list, Mat_np, Mat_pf):
```
```suggestion
        TransmissionMatrix.from_abcd(A, B, C, D, 1000)
        
def test_tmatrix_from_abcd_input_types(frequencies, Mat_list, Mat_np, Mat_pf):
```

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,193 @@
+# -------------------------
+@pytest.fixture(scope="module")
+def abcd_data_3x2():
+    """ABCD matrices with 2 frequency bins andone additional
```
```suggestion
    """ABCD matrices with 2 frequency bins and one additional
```

sikersten reviewed (changes requested):
Thanks, great work! I have some comments, but nothing too critical. I agree with @ahms5 that this can soon be merged :)

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+            The impedance data of the series impedance.
+        abcd_cshape : tuple | int, optional
+            Shape of additional channel dimensions besides (2,2).
+            Per default, no additional dimensions are created.
```
This is not how it works currently. A T-Matrix with 1D channel data for ABCD will have a (2,2) cshape. If you then access the A-property you get a FrequencyData object with cshape (1,). This is why abcd_cshape returns (1,) for Single-Channel T-Matrices and not ().

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        -------
+        idx_inf : logical array
+            An np.ndarray of logicals pointing to elements referring to
+            Zl = inf. The shape is broadcasted to self.A.freq so that it can be
```
No

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        Zl : scalar | FrequencyData
+            The load impedance data as scalar or FrequencyData. The shape must
+            match the entries of the T-matrix, i.e.
+            shape(tmat.A.freq) == shape(Zl.freq) or must be broadcastable.
```
Frequency bins must also match, thus comparing the shape not cshape

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        """Calculates the output impedance given the load impedance Zl at the
+        input.
+
+        Two-port representation::
+
+                    o---xxxxxxxxx---o
+                    |   x       x
+                    Zl  x       x      <--Zout
+                    |   x       x
+                    o---xxxxxxxxx---o
```
I would keep those.
I currently have not enough time for additional sketches for the other functions. Not sure whether this in the same style, maybe for the other function an actual PNG was required. If this is really required, I suggest adding an issue for future.

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        """Returns the transfer function (output/input) for specified
+        quantities and given load impedance.
+
+        The transfer function (TF) is the relation between an output and input
```
Yes!

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+        >>> tmat = pf.TransmissionMatrix.from_abcd(A,B,C,D, frequencies)
+
+        """
+        num_freqdata = 0
```
It makes sense to work on the raw data, since the order of dimension requires a change too

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,670 @@
+"""
+The following introduces the :py:class:`~TransmissionMatrix` class available
```
No, but T-matrix rlz, bro!

ph-schaefer commented:
Major changes since last review:
- create methods now work for frequency- dependent/independent data
   - if frequency-dependent input data => TransmissionMatrix object returned
   - otherwise 2x2 np.ndarray returned which can be multiplied to other T-Matrices
- Implemented proper slicing, now
   - only works on ABCD channel, i.e. cannot break 2x2 matrix structure
   - returns T-Matrix object
- implemented IO functionality
  - **independent testing is required**
  - I added unit test in test_io.py: `test_write_read_transmissionmatrix`
- added short doc-strings to unit tests
- added code example to `input_impedance` doc string

sikersten commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -0,0 +1,796 @@
+        refers to the "voltage-like" quantity (:math:`Q_1`) whereas the second
+        refers to the "current-like" quantity (:math:`Q_2`).
+
+        The TFs can be derived from Equation (2-1) in Reference [1]_:
```
I don't think it is necessary to introduce the abbreviation "TF" for this docstring. Could please replace it with transfer function? :)

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,301 @@
+    assert tmat.shape == (2,2)
+    npt.assert_allclose(tmat, [[1, 0],[Y, 1]], atol=1e-15)
+
+@pytest.mark.parametrize("transducer_contant",
```
```suggestion
@pytest.mark.parametrize("transducer_constant",
```

sikersten commented on tests/test_transmission_matrix.py:
```diff
@@ -0,0 +1,301 @@
+    Zin = tmat.input_impedance(Zl)
+    npt.assert_allclose(Zin.freq, Zin_expected, atol = 1e-15)
+
+@pytest.mark.parametrize("transducer_contant",
```
```suggestion
@pytest.mark.parametrize("transducer_constant",
```

sikersten commented on docs/classes/pyfar.transmission_matrix.rst:
```diff
@@ -0,0 +1,7 @@
+.. automodule:: pyfar.classes.transmission_matrix
+   :members:
+   :undoc-members:
+   :inherited-members: FrequencyData
```
This does not seem to work, but should be fixed by
```suggestion
   :inherited-members: pyfar.FrequencyData
```

sikersten reviewed (changes requested):
Thanks, only minor things left, otherwise approved :)

ph-schaefer commented on pyfar/classes/transmission_matrix.py:
```diff
@@ -81,7 +83,8 @@ class TransmissionMatrix(FrequencyData):
         """Initialize TransmissionMatrix with data, and frequencies.
 
-        This should not be used directly. Instead use :py:func:`~from_tmatrix`.
+        This should not be used directly. Instead use :py:func:`~from_tmatrix`
```
This was the intended doc string: from_tmatrix() simply calls init() and expects that data is already in the (...2,2,N) format.

sikersten replied:
Ah ok, then you need to reomve line 87.

ph-schaefer replied:
Nevermind, your original comment was correct, I oversaw line 87 :P

ahms5 reviewed (dismissed):
Thank you a lot. Seems like we need to merge dev into this pr befreo merging. Otherwise its approved.
